File: template/Template.java
public class Template {

    private abstract class WorkerTemplate {
        public final void execute() {
            startWork();
            doWork();
            finishWork();
        }

        public void startWork() {
            System.out.println("Start working");
        }

        public abstract void doWork();

        public void finishWork() {
            System.out.println("Finish working");
        }

    }

    private class HRWorker extends WorkerTemplate {
        @Override
        public void doWork() {
            System.out.println("HR worker is working");
        }
    }

    private class DeveloperWorker extends WorkerTemplate {
        @Override
        public void doWork() {
            System.out.println("Developer worker is writing code");
        }
    }

    private class ManagerWorker extends WorkerTemplate {
        @Override
        public void doWork() {
            System.out.println("Manager worker is managing");
        }
    }

    public void testTemplate() {
        WorkerTemplate worker = new HRWorker();
        worker.execute();

        worker = new DeveloperWorker();
        worker.execute();

        worker = new ManagerWorker();
        worker.execute();
    }

    public static void main(String[] args) {
        Template template = new Template();
        template.testTemplate();
    }
}

File: template/Template.ts

abstract class WorkerTemplate {
    startWork() {
        console.log("Start working");
    }
    abstract doWork();

    finishWork() {
        console.log("Finish working");
    }

    execute() {
        this.startWork();
        this.doWork();
        this.finishWork();
    }
}

class HRWorker extends WorkerTemplate {
    doWork() {
        console.log("HR worker is hiring");
    }
}

class ManagerWorker extends WorkerTemplate {
    doWork() {
        console.log("Manager worker is hiring");
    }
}

class DeveloperWorker extends WorkerTemplate {
    doWork() {
        console.log("Developer worker is writing code");
    }
}

const testWorkerTemplate = () => {
    let worker: WorkerTemplate = new HRWorker();
    worker.execute();

    worker = new DeveloperWorker();
    worker.execute();

    worker = new ManagerWorker();
    worker.execute();


}

testWorkerTemplate();